modules/javafx.base/src/main/java/com/sun/javafx/binding/ListenerManager.java

```diff
@@ -0,0 +1,174 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.binding;
+
+import java.util.Objects;
+
+import javafx.beans.InvalidationListener;
+import javafx.beans.value.ChangeListener;
+import javafx.beans.value.ObservableValue;
+
+/**
+ * Manages a single data field of type {@link Object} to store zero,
+ * one or more {@link InvalidationListener}s and {@link ChangeListener}s. This
+ * helps to minimize the storage requirements for keeping track of these
+ * listeners.<p>
+ *
+ * When there are no listeners, the field will be {@code null}. When there is
+ * only a single invalidation listener, the field will contain only that
+ * listener (change listeners are wrapped to track old value). When there are more
+ * than one listeners, the field will hold a {@link ListenerList}. It is recommended
+ * to never inspect this field directly but always use this manager to
+ * interact with it.
+ *
+ * @param <T> the type of the values
+ * @param <I> the type of the instance providing listener data
+ */
+public abstract class ListenerManager<T, I extends ObservableValue<T>> extends ListenerManagerBase<T, I> {
+
+    /**
+     * Adds an invalidation listener.
+     *
+     * @param instance the instance to which the listeners belong, cannot be {@code null}
+     * @param listener a listener to add, cannot be {@code null}
+     * @throws NullPointerException when listener is {@code null}
+     */
+    public void addListener(I instance, InvalidationListener listener) {
+        Objects.requireNonNull(listener);
+
+        instance.getValue();  // always trigger validation when adding an invalidation listener (required by tests)
+
+        Object data = getData(instance);
+
+        if (data == null) {
+            setData(instance, listener);
+        }
+        else if (data instanceof ListenerList list) {
+            list.add(listener);
+        }
+        else {
+            setData(instance, new ListenerList(data, listener));
+        }
+    }
+
+    /**
+     * Adds a change listener.
+     *
+     * @param instance the instance to which the listeners belong, cannot be {@code null}
+     * @param listener a listener to add, cannot be {@code null}
+     * @throws NullPointerException when listener is {@code null}
+     */
+    public void addListener(I instance, ChangeListener<? super T> listener) {
```

Comment:
Yes, it would be good to just put these `static` somewhere.  They don't really fit well in the manager class (as the List class which doesn't depend on manager would need to call them there), and the other way around is also odd, given that I only need to call them when *not* in "list" mode.  Still, putting them package private in the ListenerListBase could work....